### PR TITLE
embedding fuse layernorm

### PR DESCRIPTION
**Pull request by carlushuang**
add embedding fused with layernorm example and device instances.

Current limitation: embedding-dim must be equal to the template parameter ( we need to do layernorm later), and it must be multiple of `BlockSize` (256)

Currently manually implemented the `buffer_load` operation instead of using that in any version of `threadwise_copy`, the main reason is we need to accumulate the `row-index` to the base pointer, in order to support larger GPU tensor > 4G.

Below test on a gfx90a machine, with `index-length = 2048, table-length = 65536`, psudo-random initialize row index.
| emb_dim(fp16) | time(ms) | bandwidth(GB/s) |
| -------- | ------- |------- |
| 256 | 0.0249 | 168.404 |
| 512 | 0.0271 | 309.573 |
| 768 | 0.0298 | 421.556 |
| 1024 | 0.0238 | 705.809 |
| 1536 | 0.0322 | 780.381 |
| 2048 | 0.0386 | 868.236 |
| 4096 | 0.0666 | 1007.76 |
| 8192 | 0.1297 | 1034.35 |

| emb_dim(fp32) | time(ms) | bandwidth(GB/s) |
| -------- | ------- |------- |
| 256 | 0.0251 | 334.023 |
| 512 | 0.0236 | 711.555 |
| 768 | 0.0367 | 684.614 |
| 1024 | 0.0376 | 891.862 |
| 1536 | 0.0598 | 841.083 |
| 2048 | 0.0733 | 914.81 |
| 4096 | 0.1287 | 1043.09 |
| 8192 | 0.2509 | 1069.89 |

### CODE REVIEW

**Comment by asroy** [comment]
@carlushuang  Very nice code!

File: include/ck/tensor_operation/gpu/grid/gridwise_sparse_embedding3_forward_layernorm.hpp [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+        const auto thread_dim_cluster_id = thread_cluster_idx[I0];
+        const auto thread_row_cluster_id = thread_cluster_idx[I1];
+
+        const auto wave_dim_id = __builtin_amdgcn_readfirstlane(thread_dim_cluster_id / WaveSize);
+
+        const auto index_start = block_global_id * DimPerBlock + wave_dim_id * DimThreadSize;
+
+        auto threadwise_welford       = ThreadwiseWelford();
+        threadwise_welford.max_count_ = RowSubBlocks * RowVectorSize;
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_sparse_embedding3_forward_layernorm.hpp:131]: As discussion in teams, we need to deal with padding.
You could imitate the implementation from here
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/0bd6b842b96d052e03b4726ad63f8d337550cf1f/include/ck/tensor_operation/gpu/grid/gridwise_layernorm_welford_variance.hpp#L225

**Comment by asroy** [comment]
@carlushuang CI failure due to cosmetic issue, code need cleaning

image

File: include/ck/tensor_operation/gpu/grid/gridwise_sparse_embedding3_forward_layernorm.hpp (hunk 1 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/grid/gridwise_sparse_embedding3_forward_layernorm.hpp:131]: There is more work need to be able to support padding automatically like other device op. Threadwise tensor copy need to be modified. I think for now, it's ok to not support padding.

**Comment by carlushuang** [comment]
clang-tidy fixed. Currently blocked by #409 to fix a compile error

File: include/ck/utility/amd_buffer_addressing.hpp [hunk 2]
```diff
@@ -34,6 +34,21 @@ __device__ int32x4_t make_wave_buffer_resource(T* p_wave, index_t element_space_
...
 
+template <typename T>
+__device__ int32x4_t make_wave_buffer_resource_with_default_range(T* p_wave)
+{
+    BufferResource<T> wave_buffer_resource;
+
+    // wavewise base address (64 bit)
+    wave_buffer_resource.address(Number<0>{}) = const_cast<remove_cv_t<T>*>(p_wave);
+    // wavewise range (32 bit)
+    wave_buffer_resource.range(Number<2>{}) = 0xffffffff; // max possible range
```
Comment (rocking5566) [inline include/ck/utility/amd_buffer_addressing.hpp:45]: To reduce the number of function, I thought you can just modify make_wave_buffer_resource.
Add template parameter like this.
template <typename T, bool default_range=false>

Comment (carlushuang) [reply include/ck/utility/amd_buffer_addressing.hpp:45]: One thing about this approach is, I still need to pass in a `element_space_size` parameter as input. Although I can still provide this parameter with a default value, but we may result in 2 parameter both with default value, we may potentially result in unwanted behavior. So here I suggest keep it this way.

**Comment by rocking5566** [comment]
I found this implementation is embedding + 3-nary elementwise add + layernorm.
In next (future) PR, I suggest you could use tuple of input + lambda function.
If tuple size == 1, it is usual embedding layer + layernorm.
If tuple size > 1, it is embedding layer + elementwise (not only add) + layernorm.

**Comment by carlushuang** [comment]
> 

Thanks for this suggestion. Like we talked, the reason why I keep this implementation simple is because, in case of too much abstraction and result in hard to modify the code, if I need to modify the computation.
There are more to be refactored in both embedding part and layernorm part in order to have a high-quality and reusable code, so that would in next phase and depends on priority.

**Review by asroy — APPROVED** [review]

**Pull request merged**